File: Algorithms/p008_String_to_Integer_(atoi)/p008_String_to_Integer_(atoi).py
class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        i = 0
        while i<len(str) and str[i]==' ':
            i += 1
        if i<len(str):
            res = 0
            positive = True
            if str[i]=='+':
                i += 1
            elif str[i]=='-':
                positive = False
                i += 1
            while i<len(str):
                if '0'<=str[i]<='9':
                    res = res*10+int(str[i])
                    if res>2147483647:
                        return 2147483647 if positive else -2147483648
                    i += 1
                else:
                    break
            return res if positive else -res
        return 0

File: Algorithms/p008_String_to_Integer_(atoi)/p008_String_to_Integer_(atoi).cpp
class Solution {
public:
    int myAtoi(string str) {
        int i = 0;
        while(i<str.size()&&str[i]==' ') ++i;
        if(i<str.size()){
            long long res = 0;
            bool positive = 1;
            if(str[i]=='+'){
                ++i;
            }else if(str[i]=='-'){
                positive = 0;
                ++i;
            }
            while(i<str.size()){
                if(str[i]>='0'&&str[i]<='9'){
                    res = res*10+(str[i]-'0');
                    if(res>numeric_limits<int>::max()){
                        return positive ? numeric_limits<int>::max() : numeric_limits<int>::min();
                    }
                    ++i;
                }else{
                    break;
                }
            }
            return positive ? res : -res;
        }
        return 0;
    }
};
